src/Core/Managed/Shared/ActivityExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿#if !NET40
+namespace Microsoft.ApplicationInsights
+{
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+
+    internal static class ActivityExtensions
+    {
+        private const string OperationNameTag = "OperationName";
+        private static bool isInitialized = false;
+        private static bool isEnabled = false;
+
+        /// <summary>
+        /// Checks that Activity API could be called (DiagnosticSource DLL is loaded).
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)] 
```

Comment:
```NoInlining``` makes sense, but ```NoOptimization``` shouldn't be necessary. In fact, you probably want a fast, inlinable check for isEnabled. Refactor this into the fast and slow parts like this:
i.e.
```C#
// Fast, inlinable
public static bool IsActivityEnabled() => isInitialized ? isEnabled : SlowPath();

// Slow
[MethodImpl(MethodImplOptions.NoInlining)]
public static bool SlowPath()
{
    // etc.
}